netobserv/v1alpha1/flowcollector_types.go

```diff
@@ -0,0 +1,436 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package v1alpha1
+
+import (
+	ascv2 "k8s.io/api/autoscaling/v2beta2"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+const (
+	AgentIPFIX = "ipfix"
+	AgentEBPF  = "ebpf"
+)
+
+// Please notice that the FlowCollectorSpec's properties MUST redefine one of the default
+// values to force the definition of the section when it is not provided by the manifest.
+// This will cause that the remaining default fields will be set according to their definition.
+// Otherwise, omitting the sections in the manifest would lead to zero-valued properties.
+// This is a workaround for the related issue:
+// https://github.com/kubernetes-sigs/controller-tools/issues/622
+
+// FlowCollectorSpec defines the desired state of FlowCollector
+type FlowCollectorSpec struct {
+	// Important: Run "make generate" to regenerate code after modifying this file
+
+	//+kubebuilder:default:=""
+	// Namespace where NetObserv pods are deployed.
+	// If empty, the namespace of the operator is going to be used.
+	Namespace string `json:"namespace,omitempty"`
+
+	//+kubebuilder:validation:Enum=ipfix;ebpf
+	//+kubebuilder:default:=ipfix
+	// Select the flows tracing agent. Possible values are "ipfix" (default) to use
+	// the IPFIX collector, or "ebpf" to use NetObserv eBPF agent. When using IPFIX with OVN-Kubernetes
+	// CNI, NetObserv will configure OVN's IPFIX exporter. Other CNIs are not supported, they could
+	// work but necessitate manual configuration.
+	Agent string `json:"agent"`
+
+	// Settings related to IPFIX-based flow reporter when the "agent" property is set
+	// to "ipfix".
+	// +kubebuilder:default:={sampling:400}
+	IPFIX FlowCollectorIPFIX `json:"ipfix,omitempty"`
+
+	// Settings related to eBPF-based flow reporter when the "agent" property is set
+	// to "ebpf".
+	// +kubebuilder:default={imagePullPolicy:"IfNotPresent"}
+	EBPF FlowCollectorEBPF `json:"ebpf,omitempty"`
```

Comment:
This looks like a discriminated union, https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions, you should have this in a separate struct else it can become confusing for users. This will make things a bit more clear as to which fields should be specified when

```
agent:
  type: EBPF
  ebpf:
     ...
```